File: minStack.cpp
#include<iostream>
using namespace std; 

class MinStack {
private: 
    struct Node
    {
        int val; 
        int mini; 
        Node* next; 
        Node(int val,int mini,Node* next): val(val),mini(mini), next(next){}
    };
    Node* topPtr; 
public:
    MinStack() {
        topPtr = nullptr; 
    }
     
    void push(int val) {
        if (topPtr == nullptr) {
            topPtr = new Node(val,val,nullptr); 
        } else {
            int mn = min(val,topPtr->mini); 
            Node* newNode = new Node(val,mn,topPtr); 
            topPtr = newNode; 
        }
    }

    void pop() {
        Node* tmp = topPtr; 
        topPtr = topPtr->next; 
        delete tmp; 
    }
    
    int top() {
        return topPtr->val; 
    }
    
    int getMin() {
        return topPtr->mini; 
    }
};

File: minstack.py
class MinStack(object):

    def __init__(self):
        self.stack = []

    def push(self, val):
        if self.stack: 
            x = min(self.stack[-1][1], val)
            self.stack.append([val,x])
        else:
            self.stack.append([val,val])
            
        

    def pop(self):
        self.stack.pop()
        

    def top(self):
        return self.stack[-1][0]
        

    def getMin(self):
        return self.stack[-1][1]
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()